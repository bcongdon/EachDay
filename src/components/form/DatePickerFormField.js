import React from 'react'
import PropTypes from 'prop-types'
import DatePicker from 'react-datepicker'
import moment from 'moment'

import 'react-datepicker/dist/react-datepicker.css'

class DatePickerFormField extends React.Component {
  static propTypes = {
    input: PropTypes.shape({
      onChange: PropTypes.func.isRequired,
      value: PropTypes.string.isRequired
    }).isRequired,
    meta: PropTypes.shape({
      touched: PropTypes.bool,
      error: PropTypes.bool
    })
  }

  constructor (props) {
    super(props)
    this.handleChange = this.handleChange.bind(this)

    // Default to today
    this.handleChange(moment())
  }

  handleChange (date) {
    this.props.input.onChange(moment(date).format('YYYY-MM-DD'))
  }

  render () {
    const { input, meta: {touched, error} } = this.props

    return (
      <div>
        <DatePicker
          {...input}
          dateFormat='YYYY-MM-DD'
          todayButton={'Today'}
          onChange={this.handleChange}
        />
        {touched && error && <span>{error}</span>}
      </div>
    )
  }
}

export default DatePickerFormField
