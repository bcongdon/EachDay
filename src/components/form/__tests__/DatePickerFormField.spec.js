import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import DatePickerFormField from '../DatePickerFormField'

const onChange = jest.fn()

describe('DatePickerFormField component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
    Date.now = jest.fn(() => 1495984743025)
    Date.prototype.getTimezoneOffset = () => 0 // eslint-disable-line no-extend-native
  })

  it('should render correctly', () => {
    var field = mount(
      <DatePickerFormField
        input={{onChange, value: '2017-01-01'}}
        meta={{touched: false, error: ''}}
      />
    )
    expect(toJson(field)).toMatchSnapshot()
  })

  it('should set default date if none provided', () => {
    mount(
      <DatePickerFormField
        input={{onChange, value: ''}}
        meta={{touched: false, error: ''}}
      />
    )
    expect(onChange).toBeCalled()
    expect(onChange.mock.calls[0][0]).toEqual('2017-05-28')
  })

  it('should not display error if not touched', () => {
    var field = mount(
      <DatePickerFormField
        input={{onChange, value: '2017-01-01'}}
        meta={{touched: false, error: 'Warning!'}}
      />
    )
    expect(field.find('span').length).toBe(0)
    expect(toJson(field)).toMatchSnapshot()
  })

  it('should display error if touched', () => {
    var field = mount(
      <DatePickerFormField
        input={{onChange, value: '2017-01-01'}}
        meta={{touched: true, error: 'Warning!'}}
      />
    )
    expect(field.find('span').length).toBe(1)
    expect(field.find('span').first().text()).toEqual('Warning!')
    expect(toJson(field)).toMatchSnapshot()
  })
})
