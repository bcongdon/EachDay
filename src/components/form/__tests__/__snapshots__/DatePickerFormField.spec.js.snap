// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePickerFormField component should display error if touched 1`] = `
<DatePickerFormField
  input={
    Object {
      "onChange": [Function],
      "value": "2017-01-01",
    }
  }
  meta={
    Object {
      "error": "Warning!",
      "touched": true,
    }
  }
>
  <div>
    <t
      dateFormat="YYYY-MM-DD"
      dateFormatCalendar="MMMM YYYY"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      monthsShown={1}
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      popoverAttachment="top left"
      popoverTargetAttachment="bottom left"
      popoverTargetOffset="10px 0"
      selected={"2017-01-01T00:00:00.000Z"}
      tetherConstraints={
        Array [
          Object {
            "attachment": "together",
            "to": "window",
          },
        ]
      }
      todayButton="Today"
      utcOffset={-0}
      withPortal={false}
    >
      <t
        attachment="top left"
        classPrefix="react-datepicker__tether"
        constraints={
          Array [
            Object {
              "attachment": "together",
              "to": "window",
            },
          ]
        }
        renderElementTag="div"
        renderElementTo={null}
        targetAttachment="bottom left"
        targetOffset="10px 0"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            className=""
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            type="text"
            value="2017-01-01"
          />
        </div>
      </t>
    </t>
    <span>
      Warning!
    </span>
  </div>
</DatePickerFormField>
`;

exports[`DatePickerFormField component should not display error if not touched 1`] = `
<DatePickerFormField
  input={
    Object {
      "onChange": [Function],
      "value": "2017-01-01",
    }
  }
  meta={
    Object {
      "error": "Warning!",
      "touched": false,
    }
  }
>
  <div>
    <t
      dateFormat="YYYY-MM-DD"
      dateFormatCalendar="MMMM YYYY"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      monthsShown={1}
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      popoverAttachment="top left"
      popoverTargetAttachment="bottom left"
      popoverTargetOffset="10px 0"
      selected={"2017-01-01T00:00:00.000Z"}
      tetherConstraints={
        Array [
          Object {
            "attachment": "together",
            "to": "window",
          },
        ]
      }
      todayButton="Today"
      utcOffset={-0}
      withPortal={false}
    >
      <t
        attachment="top left"
        classPrefix="react-datepicker__tether"
        constraints={
          Array [
            Object {
              "attachment": "together",
              "to": "window",
            },
          ]
        }
        renderElementTag="div"
        renderElementTo={null}
        targetAttachment="bottom left"
        targetOffset="10px 0"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            className=""
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            type="text"
            value="2017-01-01"
          />
        </div>
      </t>
    </t>
  </div>
</DatePickerFormField>
`;

exports[`DatePickerFormField component should render correctly 1`] = `
<DatePickerFormField
  input={
    Object {
      "onChange": [Function],
      "value": "2017-01-01",
    }
  }
  meta={
    Object {
      "error": "",
      "touched": false,
    }
  }
>
  <div>
    <t
      dateFormat="YYYY-MM-DD"
      dateFormatCalendar="MMMM YYYY"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      monthsShown={1}
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      popoverAttachment="top left"
      popoverTargetAttachment="bottom left"
      popoverTargetOffset="10px 0"
      selected={"2017-01-01T00:00:00.000Z"}
      tetherConstraints={
        Array [
          Object {
            "attachment": "together",
            "to": "window",
          },
        ]
      }
      todayButton="Today"
      utcOffset={-0}
      withPortal={false}
    >
      <t
        attachment="top left"
        classPrefix="react-datepicker__tether"
        constraints={
          Array [
            Object {
              "attachment": "together",
              "to": "window",
            },
          ]
        }
        renderElementTag="div"
        renderElementTo={null}
        targetAttachment="bottom left"
        targetOffset="10px 0"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            className=""
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            type="text"
            value="2017-01-01"
          />
        </div>
      </t>
    </t>
  </div>
</DatePickerFormField>
`;
